Rubberduck.RibbonDispatcher/Sample Ribbons/RibbonLibrary/src/RibbonCallbacksAdapter.bas

```diff
@@ -0,0 +1,226 @@
+Sub GetItemCount(ByVal Control As IRibbonControl, ByRef Count)
+    On Error GoTo EH
+    Count = ViewModelFor(Control).GetItemCount(Control)
+XT: Exit Sub
+EH: If Err.Number <> 91 Then DisplayError Err, ModuleName & "GetItemCount", Control.Id
+    Resume XT
+End Sub
+Sub GetSelectedItemID(ByVal Control As IRibbonControl, ByRef Id)
+    On Error GoTo EH
+    Id = ViewModelFor(Control).GetSelectedItemID(Control)
+XT: Exit Sub
+EH: If Err.Number <> 91 Then DisplayError Err, ModuleName & "GetSelectedItemID", Control.Id
+    Resume XT
+End Sub
+Sub GetSelectedItemIndex(ByVal Control As IRibbonControl, ByRef Index)
+    On Error GoTo EH
+    Index = ViewModelFor(Control).GetSelectedItemIndex(Control)
+XT: Exit Sub
+EH: If Err.Number <> 91 Then DisplayError Err, ModuleName & "GetSelectedItemIndex", Control.Id
+    Resume XT
+End Sub
+''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+'''               Selectable Items - DropDown- & ComboBox-Items
+''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+Sub GetItemID(ByVal Control As IRibbonControl, ByVal Index As Integer, ByRef Id)
+    On Error GoTo EH
+    Id = ViewModelFor(Control).GetItemID(Control, Index)
+XT: Exit Sub
+EH: If Err.Number <> 91 Then DisplayError Err, ModuleName & "GetItemID", Control.Id
+    Resume XT
+End Sub
+Sub GetItemImage(ByVal Control As IRibbonControl, ByVal Index As Integer, ByRef Image)
+    On Error GoTo EH
+    Set Image = ViewModelFor(Control).GetItemImage(Control, Index)
+XT: Exit Sub
+EH: If Err.Number <> 91 Then Image = "MacroSecurity" ' DisplayError Err, ModuleName & "GetItemImage"
+    Resume XT
+End Sub
+Sub GetItemLabel(ByVal Control As IRibbonControl, ByVal Index As Integer, ByRef Label)
+    On Error GoTo EH
+    Label = ViewModelFor(Control).GetItemLabel(Control, Index)
+XT: Exit Sub
+EH: If Err.Number <> 91 Then DisplayError Err, ModuleName & "GetItemLabel", Control.Id
+    Resume XT
+End Sub
+Sub GetItemScreenTip(ByVal Control As IRibbonControl, ByVal Index As Integer, ByRef ScreenTip)
+    On Error GoTo EH
+    ScreenTip = ViewModelFor(Control).GetItemScreenTip(Control, Index)
+XT: Exit Sub
+EH: If Err.Number <> 91 Then DisplayError Err, ModuleName & "GetItemScreenTip", Control.Id
+    Resume XT
+End Sub
+Sub GetItemSuperTip(ByVal Control As IRibbonControl, ByVal Index As Integer, ByRef SuperTip)
+    On Error GoTo EH
+    SuperTip = ViewModelFor(Control).GetItemSuperTip(Control, Index)
+XT: Exit Sub
+EH: If Err.Number <> 91 Then DisplayError Err, ModuleName & "GetItemSuperTip", Control.Id
+    Resume XT
+End Sub
+Sub OnActionDropDown(ByVal Control As IRibbonControl, ByVal SelectedId As String, ByVal SelectedIndex As Integer)
+    On Error GoTo EH
+    ViewModelFor(Control).OnActionDropDown Control, SelectedId, SelectedIndex
+XT: Exit Sub
+EH: If Err.Number <> 91 Then DisplayError Err, ModuleName & "OnActionDropDown", Control.Id
+    Resume XT
+End Sub
```

Comment:
I don't see any error handling or line label that's warranted in this module. If possible I'd like the `ByRef` out parameters to more explicitly convey their nature with their name - I usually use an `out` prefix for that. Member access modifiers should be explicit, too.

Reply:
The callback signatures are published by MS. I believe that messing with the parameter names is simply a recipe for confusion all around. 

The ribbon has a strong propensity to fail, silently and fatally, at the slightest excuse - leaving one to guess at what might have gone wrong. many of these failures occur deep in the EXCEL code itself, though being triggered by quite minor specification errors. One goal is to eliminate as many of these silent fatal failures as possible, by always either giving the Ribbon something it can at least not choke on, and then signalling the error somehow to the user/developer. I believe I have made good strides in this direction with this library.  Adding all of this error handling to the Ribbon Dispatcher, which is the piece that sits directly in contact with Excel, is part of that effort.

Reply:
But error 91 can easily be avoided with the introduction of a local variable, and any other error gets rethrown anyway. No need.

Reply:
Other than the "Part 3 of 3" article, these signatures aren't officially published anywhere, but as with any signature, unless arguments are passed by name (which I don't think you can do from a non-VBA language, to VBA), then the only parts of the signature that matter are the return type of the routine (if any), and the order and types of the arguments, not the names of the arguments themselves. There's nothing to stop me calling my `GetEnabled` callback `FetchEnabled` or `FizzTheBuzz`.

In any event, you've changed the capitalization of the parameter names, and highlighted one of the problems of the documented parameter names. `control As IRibbonControl` will conflict with declarations of `MSForms.Control` is any project that has a UserForm - i.e. the capitalization will be forced one way or the other.